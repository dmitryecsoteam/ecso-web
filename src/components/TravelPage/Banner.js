import React from 'react';
import classNames from 'classnames';
import TrackVisibility from 'react-on-screen';

export default ({ linkTo, backgroundImage, textMain, textSecondary, textButton }) => {

    const BannerVisible = ({ isVisible }) => {
        const className = classNames('banner__container', { 'banner__container--animation': isVisible });
        return (
            <div className={className}>
                <a
                    className="banner__link"
                    href={linkTo}
                >
                    <div
                        className="banner__image"
                        style={{ backgroundImage: `url(${backgroundImage})` }}
                    >
                    </div>
                    <span className="banner__text banner__text--main">{textMain}</span>
                    <span className="banner__text banner__text--secondary">{textSecondary}</span>
                    <div className="banner__button">{textButton}</div>
                </a>
            </div>
        );
    };

    return (
        <TrackVisibility once className="banner">
            <BannerVisible />
        </TrackVisibility>
    );
}