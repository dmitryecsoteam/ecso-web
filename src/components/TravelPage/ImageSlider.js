import React from 'react';
import loadImage from '../../utils/loadImage';
import '@babel/polyfill';

export default class ImageSlider extends React.Component {

    // Put first slide in initial state, so it could be displayed as soon as it would be loaded
    // Then load all next slides in componentDidMount
    state = {
        images: [`/images/${this.props.name}-${this.props.country}/slide1.jpg`],
        index: 0
    }

    // ref to ImageSlider div container. To control css classes and animation
    divRef = React.createRef();

    // Interval property for setInterval and clearInterval
    autoSlide = 0;

    async componentDidMount() {

        // add animation
        this.divRef.current.classList.add('image-slider__animation');

        const images = this.state.images.slice();
        let i = 2;
        let loadNextImage = true;
        while (loadNextImage) {

            // load next image
            loadNextImage = await loadImage(`/images/${this.props.name}-${this.props.country}/slide${i}.jpg`);

            // TRUE - image exists and loaded in cache, add this URL to images array and increment counter
            // FALSE - image doesn't exist, leave while loop
            if (loadNextImage) {
                images.push(`/images/${this.props.name}-${this.props.country}/slide${i}.jpg`);
                i++;
            }
        }

        // update state to rerender component
        this.setState({
            images
        });

        if (images.length > 1) {
            // Set interval to auto-slide when there are more than one image
            this.autoSlide = setInterval(this.slideRight, this.props.interval);
        }
    }

    componentDidUpdate() {

        if (this.divRef.current) {

            // HTMLElement's property reference triggers DOM reflow
            // without this animation wouldn't work
            // https://css-tricks.com/restart-css-animation/#article-header-id-0
            this.divRef.current.offsetWidth;

            // rerun animation on new image
            this.divRef.current.classList.add('image-slider__animation');

        }
    }

    componentWillUnmount() {
        clearInterval(this.autoSlide);
    }

    slideAndResetInterval = (slide, id) => {

        // remove auto-slide interval
        clearInterval(this.autoSlide);

        // change background image
        slide(id);

        // set new auto-slide interval
        this.autoSlide = setInterval(this.slideRight, this.props.interval);
    }

    slideTo = (index) => {

        // remove animation
        this.divRef.current.classList.remove('image-slider__animation');

        this.setState({
            index
        });
    }

    slideRight = () => {

        // rerender with next background image from images array
        // if current background image is the last in array of images, than move to the start
        if (this.state.index === this.state.images.length - 1) this.slideTo(0);
        else this.slideTo(this.state.index + 1);

    }

    slideLeft = () => {

        // rerender with previous background image from images array
        // if current background image is the first one, than move to the last one
        if (this.state.index === 0) this.slideTo(this.state.images.length - 1);
        else this.slideTo(this.state.index - 1);

    }

    onLeftClick = () => {

        // slide to previous image
        this.slideAndResetInterval(this.slideLeft);

    }

    onRightClick = () => {

        // slide to next image
        this.slideAndResetInterval(this.slideRight);

    }

    onDotClick = (event) => {

        // slide to choosed image
        this.slideAndResetInterval(this.slideTo, +event.target.dataset.id);

    }

    render() {

        const { index } = this.state;

        const { images } = this.state;

        const dots = [];
        for (let i = 0; i < images.length; i++) {
            let className = "image-slider__dot";
            if (i === index) {
                className += " image-slider__dot--active"
            }
            dots.push(<span key={i} data-id={i} className={className} onClick={this.onDotClick}></span>)
        }

        // const containerClassNames = classNames('image-slider__container',
        //     {
        //         'image-slider__container--empty': images.length === 0
        //     }
        // );

        return (
            <div>
                {images.length !== 0 && <div className="image-slider__container">
                    <div
                        ref={this.divRef}
                        className="image-slider__current-slide"
                        style={{ backgroundImage: `url(${images[index]})` }}
                    >
                    </div>
                    <div className="image-slider__button image-slider__button--left" onClick={this.onLeftClick}>&#10094;</div>
                    <div className="image-slider__button image-slider__button--right" onClick={this.onRightClick}>&#10095;</div>
                    <div className="image-slider__dots-container">{dots}</div>
                </div>}
            </div>
        );
    }
}