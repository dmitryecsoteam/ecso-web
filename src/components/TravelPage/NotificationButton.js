import React from 'react';
import { ClipLoader } from 'halogenium';

import { withSession } from '../../auth/session';
import { client } from '../../clientGraphQL/client';

import { GET_NOTIFICATIONS } from '../../queries/queries';
import { ADD_NOTIFICATION, DELETE_NOTIFICATION } from '../../queries/mutations';
import NotificationIcon from '../../images/icons/notification.png';
import DeleteNotificationIcon from '../../images/icons/delete-notification.png';




export class NotificationButton extends React.Component {

    state = {
        // If current travel is in user's notification list
        isNotification: false,
        loadingNotification: false
    }

    componentDidMount() {
        this.updateNotification();
    }

    componentDidUpdate = async (prevProps) => {
        // If user changed and it's not null - refetch notifications
        if (this.props.user && prevProps.user !== this.props.user) {
            this.updateNotification();
        }
    }

    updateNotification = async () => {
        // if user is not null get list of notifications
        if (this.props.user) {
            const { data } = await client.query({
                query: GET_NOTIFICATIONS
            });

            // Set isNotification to TRUE if travelId is in user notifications array
            this.setState({
                isNotification: data.getNotifications.some(notification => notification.travelId === this.props.travelId)
            });
        }
    }

    handleAddNotification = async () => {
        // Protect from double-click
        if (!this.state.loadingNotification) {

            this.setState({
                loadingNotification: true
            });

            await client.mutate({
                mutation: ADD_NOTIFICATION,
                variables: { id: this.props.travelId }
            });

            this.setState({
                isNotification: true,
                loadingNotification: false
            });

        }
    }

    handleDeleteNotification = async () => {
        // Protect from double-click
        if (!this.state.loadingNotification) {

            this.setState({
                loadingNotification: true
            });

            await client.mutate({
                mutation: DELETE_NOTIFICATION,
                variables: { id: this.props.travelId }
            });

            this.setState({
                isNotification: false,
                loadingNotification: false
            });

        }
    }

    render() {
        const { user, travelId } = this.props;

        const spinnerNotification = (
            <div className="add-notification__spinner-container">
                <ClipLoader className="add-notification__spinner" color="#fff" size="26px" />
            </div>
        );

        const addNotificationButton = (
            <button
                className="add-notification__button add-notification__button--big"
                disabled={this.state.loadingNotification}
                onClick={this.handleAddNotification}
            >
                <img className="travel__notification-icon" src={NotificationIcon} />
                Notify me about price changes
            </button>
        );

        const deleteNotificationButton = (
            <button
                className="add-notification__button add-notification__button--big"
                disabled={this.state.loadingNotification}
                onClick={this.handleDeleteNotification}
            >
                <img className="travel__notification-icon" src={DeleteNotificationIcon} />
                Delete notification
            </button>
        );

        // render button only if user is signed in
        return (
            <React.Fragment>
                {user &&
                    <div className="travel__notification-button">
                        {this.state.isNotification ? deleteNotificationButton : addNotificationButton}
                    </div>}
            </React.Fragment>

        )
    }
}

export default withSession(NotificationButton);