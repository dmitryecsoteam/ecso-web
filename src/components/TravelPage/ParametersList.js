import React from 'react';
import { defaultState } from '../../store/ReduxStore';
import classNames from 'classnames';
import TrackVisibility from 'react-on-screen';


export default ({ destination }) => {

    const parameters = [];
    for (let parameter in defaultState.searchForm.parametersValue) {

        const dots = [];
        for (let i = 1; i <= 5; i++) {
            const className = classNames(
                'parameters-list__dot',
                { 'parameters-list__dot--active': i <= destination[parameter.toLowerCase() + 'Rating'] }
            );
            dots.push(<span key={i} className={className}></span>);
        }

        const ParameterHeader = ({ isVisible }) => {
            const className = classNames('parameters-list__item-header', { 'parameters-list__item-header--animation': isVisible });
            return <div className={className} >
                <span className="parameters-list__label">{parameter}</span>
                <div className="parameters-list__dots-container">{dots}</div>
            </div>
        };

        parameters.push(<div key={parameter} className="parameters-list__item" >

            <TrackVisibility once>
                <ParameterHeader />
            </TrackVisibility>

            <div className="parameters-list__description">{destination[parameter.toLowerCase() + 'Description']}</div>
        </div>);
    }

    return <div className="parameters-list__container">{parameters}</div>
}