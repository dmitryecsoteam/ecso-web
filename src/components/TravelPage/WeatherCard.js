import React from 'react';
import moment from 'moment';

export default ({ tempMin, tempMax, conditionText, conditionImage, date }) => {
    moment.locale('en');
    return (
        <div className="weather-card__container">
            <div className="weather-card__icon">
                <img src={conditionImage} />
            </div>
            <p className="weather-card__title">Average weather on <nobr><b>{moment(date).format('DD MMMM YYYY')}</b></nobr></p>
            <p className="weather-card__temperature">{tempMin}°...{tempMax}°</p>
            <p className="weather-card__condition">{conditionText}</p>
        </div>
    )
}