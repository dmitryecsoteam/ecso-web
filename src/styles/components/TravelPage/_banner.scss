.banner {
    width: 100%;
    max-width: $content-width;
}

.banner__container {
    position: relative;
    height: 21.2rem;
    overflow: hidden;
    text-align: center;
    padding-top: $s-size;
    opacity: 0;
    transition: all .5s;
    transform: scale(.5,.5);
}

.banner__container--animation {
    opacity: 1;
    transform: scale(1,1);
}

.banner__link {
    text-decoration: none;
}

.banner__image {
    position: absolute;
    height: 100%;
    width: 100%;
    background: no-repeat center;
    z-index: -1;
}

.banner__image::before {
    content: "";
    display: block;
    position: absolute;
    height: 100%;
    width: 100%;
    background-color: black;
    opacity: 0.5;
  }

.banner__text-container {
    position: absolute;
    top: 50%;
    left: 50%;
}

.banner__text {
    color: white;
    display: block;
}

.banner__text--main {
   margin-top: 3.4rem;
   font-size: $l-size;
   font-weight: 500;
}

.banner__text--secondary {
    margin-top: $m-size;
}

.banner__button {
    background-color: white;
    color: $light-black;
    border-radius: $xs-size;
    padding: $m-size $s-size;
    margin-top: $s-size;
    display: inline-block;
    font-size: 1.6rem;
}

.banner__button:hover {
    background-color: lighten($light-grey, 15%);
}

@media (max-width: $desktop-breakpoint) {
    .banner__container {
        height: 15.4rem;
        padding-top: 0;
        margin-bottom: $s-size;
    }

    .banner__text--main {
        margin-top: $m-size;
        font-size: 2.8rem;
    }

    .banner__button {
        padding: $s-size;
    }
}