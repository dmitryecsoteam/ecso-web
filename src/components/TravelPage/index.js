import React from 'react';
import { Query } from 'react-apollo';
import { TRAVELS_SEARCH_FULL } from '../../queries/queries';
import Header from '../Header';
import ImageSlider from './ImageSlider';

import { client } from '../../clientGraphQL/client';

export default ({ match }) => {
    const { _id } = match.params;

    return <Query
        query={TRAVELS_SEARCH_FULL}
        variables={{ _id }}
    >
    {({ loading, error, data }) => {
        if (loading) return <p>Loading</p>;

        console.log(data)
        console.log(error)

        const { travelFull } = data;
        const { destination } = travelFull;

        return <div>
        <Header title={`${destination.nameEn}, ${destination.countryEn}`}/>
        <ImageSlider 
            interval={5000}
            name={destination.nameEn}
            country={destination.countryEn}
        />
            <h1>{`${destination.nameEn}, ${destination.countryEn}`}</h1>
        </div>
    }}
    </Query>
};