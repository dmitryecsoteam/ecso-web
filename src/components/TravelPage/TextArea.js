import React from 'react';
import WeatherCard from './WeatherCard';

export default ({ cityDescription, population, foundingDate, weatherTempMin, weatherTempMax, conditionText, conditionImage, date }) => {
    return (
        <div className="text-area__leadtext-container">
            <p className="text-area__leadtext-item">{cityDescription}</p>
            <p className="text-area__leadtext-population">Average population: {population}</p>
            <p className="text-area__leadtext-date">Founding date: {foundingDate ? foundingDate : 'N/A'}</p>
            <WeatherCard
                tempMin={weatherTempMin}
                tempMax={weatherTempMax}
                conditionText={conditionText}
                conditionImage={conditionImage}
                date={date}
            />
        </div>
    );
}