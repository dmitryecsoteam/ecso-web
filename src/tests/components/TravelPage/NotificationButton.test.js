import React from 'react';
import { shallow } from 'enzyme';
import { NotificationButton } from '../../../components/TravelPage/NotificationButton';

jest.mock('../../../clientGraphQL/client.js');

const waitForAsync = () => new Promise(resolve => setImmediate(resolve));

test('should render NotificationButton for signed user - Notify me about price changes', async () => {
    const wrapper = shallow(<NotificationButton user={{ email: 'test@mail.com', name: 'b' }} travelId="Not in notifications" />);
    await waitForAsync();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state('isNotification')).toBeFalsy();
});

test('should render NotificationButton for signed user - Delete notification', async () => {
    const wrapper = shallow(<NotificationButton user={{ email: 'test@mail.com', name: 'b' }} travelId="In notifications" />);
    await waitForAsync();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state('isNotification')).toBeTruthy();
});

test('should not render NotificationButton if user is null', () => {
    const wrapper = shallow(<NotificationButton user={null} travelId="test" />);

    expect(wrapper).toMatchSnapshot();
});