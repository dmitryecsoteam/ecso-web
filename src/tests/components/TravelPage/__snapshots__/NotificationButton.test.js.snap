// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render NotificationButton if user is null 1`] = `<Fragment />`;

exports[`should render NotificationButton for signed user - Delete notification 1`] = `
<Fragment>
  <div
    className="travel__notification-button"
  >
    <button
      className="add-notification__button add-notification__button--big"
      disabled={false}
      onClick={[Function]}
    >
      <img
        className="travel__notification-icon"
        src="test-file-stub"
      />
      Delete notification
    </button>
  </div>
</Fragment>
`;

exports[`should render NotificationButton for signed user - Notify me about price changes 1`] = `
<Fragment>
  <div
    className="travel__notification-button"
  >
    <button
      className="add-notification__button add-notification__button--big"
      disabled={false}
      onClick={[Function]}
    >
      <img
        className="travel__notification-icon"
        src="test-file-stub"
      />
      Notify me about price changes
    </button>
  </div>
</Fragment>
`;
